### ISSUE TITLE
[Question] hv.Dataset default construction from xr.DataArray 

### ISSUE THREAD

**Issue by cocoaaa** [pyviz/holoviews#3749]
Hi, I'm trying to understand how the `dims` and `coords` attributes from xr.DataArray object gets propagated to the hv.Dataset object when we don't specify any `kdims` and `vdims`, but I wasn't able to find explicit documentation on this. Based on my experiments, here is what I got:

- First, I have my data as xr.DataArray:
![image](https://user-images.githubusercontent.com/8037983/58673446-4311f600-8300-11e9-9675-a7b784a5cc38.png)

- Note its `dims` and `coords` attributes:
![image](https://user-images.githubusercontent.com/8037983/58673495-69379600-8300-11e9-8f0c-ecec792fd727.png)

- Now, I pass it to hv.Dataset without specifying `kdims` and `ydims`:
![image](https://user-images.githubusercontent.com/8037983/58673544-a1d76f80-8300-11e9-8845-543ae1ba5115.png)

- Note that the `time` dimension's `coords` is not correctly propagated:
![image](https://user-images.githubusercontent.com/8037983/58673564-b74c9980-8300-11e9-8e6f-badf05269cc4.png)

- Whereas the `y` dimension's metadata seem to have correctly propagated:
![image](https://user-images.githubusercontent.com/8037983/58673655-2a561000-8301-11e9-85d2-c3c34672abf8.png)
![image](https://user-images.githubusercontent.com/8037983/58673661-3510a500-8301-11e9-9f7d-9f3fa69a06c9.png)

- 'X' dimension also gets propagated as expected:
![image](https://user-images.githubusercontent.com/8037983/58673684-57a2be00-8301-11e9-96d5-37c9b21b4972.png)

Could you help me understand where does this difference arise from?  In general, I'm quite confused how the default constructor of hv.Element (ie. hv.Element(data, kdims=None, vdims=None) sets the kdims and vdims of the resulting Element instance. I'm trying to understand how it behaves differently when the data is unlabelled type (eg. numpy.ndarray) vs. when the data is labelled (eg. xarray.DataArray or pd.DataFrame). I'd also like to know if you could point me to where I can read upon this.  Thank you!

**Comment by philippjfr** [comment]
>Note that the time dimension's coords is not correctly propagated:

Dimensions generally only store **metadata** about the dimension, while xarray stores both metadata and data (the actual coordinate values) on the coordinates. This is where this difference is coming from, duplicating the **data** that's already inside the xarray datastructure would be very inefficient.

>In general, I'm quite confused about how the default constructor of hv.Element (ie. hv.Element(data, kdims=None, vdims=None) sets the kdims and vdims of the resulting Element instance.

This is somewhat dependent on the datatype, for unlabelled datastructures like numpy arrays it assumes the data conforms to the shape implied by the default dimensions of the element. For labelled datastructures like xarrays it will try to infer what you want. In the case of a pandas dataframe it will also try to use all the columns that are available while also respecting the dimensionality of an element (i.e. a Points element has two key dimensions so it will by default assume that the first two columns are the kdims and the remainder the vdims.

We should certainly do better about documenting this behavior.

**Comment by cocoaaa** [comment]
> This is somewhat dependent on the datatype, for unlabelled datastructures like numpy arrays it assumes the data conforms to the shape implied by the default dimensions of the element. For labeled data structures like xarrays, it will try to infer what you want. In the case of a pandas dataframe, it will also try to use all the columns that are available while also respecting the dimensionality of an element (i.e. a Points element has two key dimensions so it will by default assume that the first two columns are the kdims and the remainder the vdims.

Thank you for your clarification. 
It really helps to know that for unlabelled datastructure, holoviews' Element constructor assumes that the data conforms to the **target** Element's default dimensions! It's been a major question for me.  I think it would help a lot of people to have this explanation included in the documentation, eg. to `Getting Started -> Gridded Datasets.  In that [tutorial](https://is.gd/biXmBh), I couldn't understand why the order of `kdims` arguments needed be `['Time', 'x','y']` ( for the `hv.Dataset` constructor with np.array as the input data), since this order is different from numpy's dimensionality order (often dim0 is height, dim1 is width, dim2 is nChannels).

![image](https://user-images.githubusercontent.com/8037983/59568251-21697c00-902d-11e9-9679-a35c2d901a30.png)

Now that I'm more familiar with netCDF data model and xarray, I can see where this is coming from, but I have to say it was still quite confusing, espeically with hv.Dataset constructor which has a more flexibility in defining kdims and vdims (versus hv.Points which has a fixed number of kdims accepted). To elaborate on this,
I would expect something like this to work with a 3dimensional np.array:
```python
h,w,nc = 20,20,3
np.random.seed(0)
data = np.random.randn(h,w,nc)

# to hv.Dataset
# doesn't work as expected
ds = hv.Dataset(data, kdims=['dim0','dim1'],vdims=['r','g','b'])
```
This is because I'm expecting the `kdims` arguments to specify the mapping **from** the dimensionality of the `data`'s datastructure (np.array) to dimensions that will be used in the hv.Dataset (rather than conforming to what's expected for the downstream hv.Element constructors, ie.hv.Dataset -> hv.Image...). So, it would be great to have some reference as to when `kdims` arguments indicate the mapping from the input data to the hv.element containers versus the other way.  

Hope my attempt to explain this confusion itself is not too confusing:) Thanks!

**Comment by poplarShift** [comment]
Well if `data` is a 3D numpy array then you cannot declare more than 3 different dimensions, be they kdims or vdims, so I wouldn't intuitively expect any of your two proposed alternatives to work. I am not sure what you mean by "already existing data input" vs "expected dimensions of the elements". Every element has a number of ways to construct them. But the whole business of declaring dimensions is about _annotating the data itself_ and from there on holoviews will tell you what kind of visualizations it thinks are compatible. E.g. if your data have two kdims, you can't visualize it as a Scatter element, because Scatter is only for cases where you have one kdim and one or more vdims.

In other words, on a purely conceptual level, passing the dimensions is *external* information about the data, not in any way related to what you want to do with them. That is one of the key distinctions between imperative and declarative visualization.

Hope this helped.
